rust/hyperlane-base/src/db/rocks/hyperlane_db.rs

```diff
@@ -63,98 +66,50 @@ impl HyperlaneDB {
-            }
-        }
-
-        self.store_message(message)
-    }
-
     /// Store a raw committed message
     ///
     /// Keys --> Values:
     /// - `nonce` --> `id`
     /// - `id` --> `message`
-    pub fn store_message(&self, message: &HyperlaneMessage) -> Result<()> {
-        let id = message.id();
-
-        debug!(msg=?message, "Storing new message in db",);
-        self.store_message_id(message.nonce, message.destination, id)?;
-        self.store_keyed_encodable(MESSAGE, &id, message)?;
-        Ok(())
-    }
-
-    /// Store the latest known nonce
-    ///
-    /// Key --> value: `LATEST_NONCE` --> `nonce`
-    pub fn update_latest_nonce(&self, nonce: u32) -> Result<()> {
-        if let Ok(Some(n)) = self.retrieve_latest_nonce() {
-            if nonce <= n {
-                return Ok(());
-            }
-        }
-        self.store_encodable("", LATEST_NONCE, &nonce)
-    }
-
-    /// Retrieve the highest known nonce
-    pub fn retrieve_latest_nonce(&self) -> Result<Option<u32>> {
-        self.retrieve_decodable("", LATEST_NONCE)
-    }
-
-    /// Store the latest known nonce for a destination
-    ///
-    /// Key --> value: `destination` --> `nonce`
-    pub fn update_latest_nonce_for_destination(&self, destination: u32, nonce: u32) -> Result<()> {
-        if let Ok(Some(n)) = self.retrieve_latest_nonce_for_destination(destination) {
-            if nonce <= n {
-                return Ok(());
-            }
+    /// - `nonce` --> `dispatched block number`
+    fn store_message(
+        &self,
+        message: &HyperlaneMessage,
+        dispatched_block_number: u64,
+    ) -> DbResult<bool> {
+        if let Ok(Some(_)) = self.message_id_by_nonce(message.nonce) {
+            trace!(msg=?message, "Message already stored in db");
+            return Ok(false);
         }
-        self.store_keyed_encodable(LATEST_NONCE_FOR_DESTINATION, &destination, &nonce)
-    }
 
-    /// Retrieve the highest known nonce for a destination
-    pub fn retrieve_latest_nonce_for_destination(&self, destination: u32) -> Result<Option<u32>> {
-        self.retrieve_keyed_decodable(LATEST_NONCE_FOR_DESTINATION, &destination)
-    }
+        let id = message.id();
+        debug!(msg=?message, "Storing new message in db",);
 
-    /// Store the message id keyed by nonce
-    fn store_message_id(&self, nonce: u32, destination: u32, id: H256) -> Result<()> {
-        debug!(nonce, ?id, "storing leaf hash keyed by index");
-        self.store_keyed_encodable(MESSAGE_ID, &nonce, &id)?;
-        self.update_latest_nonce(nonce)?;
-        self.update_latest_nonce_for_destination(destination, nonce)
+        // - `id` --> `message`
+        self.store_keyed_encodable(MESSAGE, &id, message)?;
+        // - `nonce` --> `id`
+        self.store_keyed_encodable(MESSAGE_ID, &message.nonce, &id)?;
+        // - `nonce` --> `dispatched block number`
+        self.store_keyed_encodable(
```

Comment:
Iiuc if people don't wipe the relayer db before running this new version, all messages that have already been indexed won't have an entry for this. Is that ok?

Or is the assumption that this will roll out with the omniscient relayer stuff, which would mean that all messages are re-indexed anyways? (<-- i don't know if this is true, I didn't read the PR)

Reply:
It's kind of okay? The cursor currently uses this entry to detect the presence of a message, which means it will attempt to reindex all messages, which I suppose isn't great.

I think we can work around this by having it instead get messages by nonce, and if present, attempt to grab the corresponding block number. If that's found, it will use that block number to fast forward, but otherwise it will fast forward the nonce only (and not the block).

Reply:
Okay, this is what I did, it does result in an extra DB read but it's technically more correct I guess

Though I do actually assume that the relayers will need to reindex anyways, and the scraper would have been able to fast forward anyways (because it already stores the dispatched block number).

Lmk wyt